### PR TITLE
CUDA libs test: Report the absolute path of the loaded libcuda.so on Linux, + other improvements

### PR DESCRIPTION
**Pull request by gmarkall**
This PR improves the CUDA library test function. The most important change is to report the absolute path of the loaded `libcuda.so` on Linux. 

This motivation is as follows: various driver-related issues have been reported by WSL2 users (see e.g. #9032, #7104, https://github.com/rapidsai/cudf/issues/13388, other questions on Gitter, Slacks, Discourse, etc.) and it is almost always due to a Linux (i.e. not- WSL2) driver being installed in a WSL2 system. Printing the absolute path gives the version number of the driver, which can be looked up to determine whether it was a version intended for "native" Linux, or the correct WSL2 driver.

There are also some minor changes to the tests and output to improve consistency and readability - see individual commit messages for details. 

With these changes, the output of the test function (identically reported by the Numba sysinfo tool, `numba -s`) is, for example:

```
$ python -c "from numba import cuda; cuda.cudadrv.libs.test()"
Finding driver from candidates:
	libcuda.so
	libcuda.so.1
	/usr/lib/libcuda.so
	/usr/lib/libcuda.so.1
	/usr/lib64/libcuda.so
	/usr/lib64/libcuda.so.1
Using loader <class 'ctypes.CDLL'>
	Trying to load driver...	ok
		Loaded from libcuda.so
	Mapped libcuda.so paths:
		/usr/lib/x86_64-linux-gnu/libcuda.so.530.30.02
Finding nvvm from Conda environment
	Located at /home/gmarkall/mambaforge/envs/numbadev/lib/libnvvm.so.4.0.0
	Trying to open library...	ok
Finding cudart from Conda environment
	Located at /home/gmarkall/mambaforge/envs/numbadev/lib/libcudart.so.11.8.89
	Trying to open library...	ok
Finding cudadevrt from Conda environment
	Located at /home/gmarkall/mambaforge/envs/numbadev/lib/libcudadevrt.a
	Checking library...	ok
Finding libdevice from Conda environment
	Located at /home/gmarkall/mambaforge/envs/numbadev/lib/libdevice.10.bc
	Checking library...	ok
```

I would suggest that each commit can be reviewed individually - the changes in each are independent, and commit messages contain further details.

### CODE REVIEW

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
gpuci run tests

**Review by bdice — COMMENTED** [review]
I appreciate this — I’ve helped field a lot of questions from users that boil down to this core problem with WSL2 driver installation. I believe this will help with diagnosis.

The changes look fine to me.

File: numba/cuda/cudadrv/libs.py [hunk 1]
```diff
@@ -80,67 +79,97 @@ def _get_source_variable(lib, static=False):
...
+    # look up whether the driver was intended for "native" Linux.
+    if sys.platform == 'linux' and not failed:
+        pid = os.getpid()
+        mapsfile = f'/proc/{pid}/maps'
+        try:
+            with open(mapsfile) as f:
+                maps = f.read()
+        # It's difficult to predict all that might go wrong reading the maps
+        # file - in case various error conditions ensue (the file is not found,
+        # not readable, etc.) we use OSError to hopefully catch any of them.
```
Comment (stuartarchibald) [inline numba/cuda/cudadrv/libs.py:116]: Another option, if this approach does end up to be more problematic than anticipated, would perhaps be to use `dl_iterate_phdr`.

File: docs/upcoming_changes/9034.cuda.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Report absolute path of libcuda.so on Linux
+===========================================
+
+`numba -s` now reported the absolute path to libcuda.so on Linux, to aid
```
Comment (stuartarchibald) [inline docs/upcoming_changes/9034.cuda.rst:4]:
```suggestion
``numba -s`` now reports the absolute path to ``libcuda.so`` on Linux, to aid
```
think RST needs double backticks for code statements? (and line wrap)

File: numba/cuda/cudadrv/libs.py [hunk 3]
```diff
@@ -80,67 +79,97 @@ def _get_source_variable(lib, static=False):
...
+        print('\tTrying to load driver', end='...')
         dll, path = load_driver(dlloader, candidates)
-        print(f'\tok, loaded from {path}')
+        print('\tok')
+        print(f'\t\tLoaded from {path}')
     except CudaSupportError as e:
         print(f'\tERROR: failed to open driver: {e}')
         failed = True
 
+    # Find the absolute location of the driver on Linux. Various driver-related
```
Comment (stuartarchibald) [inline numba/cuda/cudadrv/libs.py:102]: The absolute path may contain information that users might not want to share, however the point of this diagnostic is that the absolute path is reported. The `numba -s` tool has a large printed warning at the end of its output with regards to checking and redacting the output before sharing which I imagine covers this.

File: numba/cuda/cudadrv/libs.py [hunk 4]
```diff
@@ -80,67 +79,97 @@ def _get_source_variable(lib, static=False):
...
+    # issues have been reported by WSL2 users, and it is almost always due to a
+    # Linux (i.e. not- WSL2) driver being installed in a WSL2 system.
+    # Providing the absolute location of the driver indicates its version
+    # number in the soname (e.g. "libcuda.so.530.30.02"), which can be used to
+    # look up whether the driver was intended for "native" Linux.
+    if sys.platform == 'linux' and not failed:
+        pid = os.getpid()
+        mapsfile = f'/proc/{pid}/maps'
```
Comment (stuartarchibald) [inline numba/cuda/cudadrv/libs.py:110]:
```suggestion
        mapsfile = os.path.join('proc', f'{pid}', 'maps')
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, couple of minor things to resolve else looks good. I've also tested this manually on a system with CUDA system level drivers, worked as expected.

File: docs/upcoming_changes/9034.cuda.rst (hunk 2 above)
Comment (gmarkall) [reply docs/upcoming_changes/9034.cuda.rst:4]: Ah, I forgot this was RST! (despite it being in the name)

**Review by stuartarchibald — DISMISSED** [review]
Thanks for the patch and fixes.

**Review by stuartarchibald — APPROVED** [review]
Thanks for fixing-up my previous suggestion via 98784c4, `proc` does indeed need accessing from the file system root. Patch re-approved!

**Comment by gmarkall** [comment]
gpuci run tests

**Pull request merged**